Update documentation for running tdvt tests using tabquerytoolsrv (#1003)
tableau/connector-plugin-sdk: docs/tdvt.md

   For example:
   `TAB_CLI_EXE_X64 = C:\Program Files\Tableau\Tableau 1234.1\bin\tabquerytool.exe`

> __Note:__ For setting up TDVT using Tableau Linux Server please click [here](/connector-plugin-sdk/docs/tdvt#running-tdvt-test-with-tabquerytoolsrv-in-linux-server). 

Author: Committed the suggestion.

__The agg.countd expression test and the join.null.int logical tests are failing__
Check that your database is correctly returning column nullability information in the metadata. See [Design Considerations]({{ site.baseurl }}/docs/design) for more information.

> ## Running TDVT test with tabquerytoolsrv.exe

Reviewer: The focus of this documentation should be the Linux use case. 

I think we want to fit this into the https://tableau.github.io/connector-plugin-sdk/docs/tdvt#set-up part. Windows and Mac users should continue to use the Desktop installers. Linux uses should install server and configure TAB_CLI_EXE_LINUX.

Author: Remove the windows part for the documentation.

Author: Updated the title to refect this is only for linux server.

You can run a particular TDVT test with the help of a test setup file using tabquerytoolsrv.exe. `tabquerytoolsrv.exe` is located in `C:\Program Files\Tableau\Tableau Server\packages\bin.<version_code>\tabquerytoolsrv.exe` for Windows and `/opt/tableau/tableau_server/packages/bin.<version_code>/tabquerytoolsrv` for Linux. 
 Before being able to run TDVT tests using tabquerytoolsrv make sure that: 
   - Your instance of the tableau server is running 

>    - Your server license has been activated 

Reviewer: Can you link to the tsm command documentation?

Author: Adding the link to starting and activating the license.

Author: Updated the link

Once the above steps are verified you should be able to run TDVT tests using tabquerytoolsrv by running the following command:

__For Windows__
```
 tabquerytoolsrv.exe --query-file-list D:\test\tests.txt -d D:\path_to_tds\cast_calcs.db.tds --combined --password-file D:\path_to_tds\db.password --output-dir D:\test -DLogDir=D:\test -DOverride=ProtocolServerNewLog -DLogicalQueryRewriteDisable=Funcall:RewriteConstantFuncall -DInMemoryLogicalCacheDisable -DParamFile="C:\ProgramData\Tableau\Tableau Server\data\tabsvc\config\minerva_0.<version_code>\workgroup.yml,C:\ProgramData\Tableau\Tableau Server\data\tabsvc\config\minerva_0.<version_code>\minerva.yml"
Attempting to run query...
Run query successful! Check output file
```

> __For Linux:__

Reviewer: Is this content below like a troubleshooting section? As long as TAB_CLI_EXE_LINUX or TAB_CLI_EXE_X64 is configured for tabquerytoolsrv, doesn't TDVT work the same as documented?

Author: I tried adding tabquerytoolsrv in `tdvt_override.ini` but this did not work. 
```
TAB_CLI_EXE_X64 = C:\Program Files\Tableau\Tableau Server\packages\bin.near.22.0802.1408\tabquerytoolsrv.exe
TAB_CLI_EXE_MAC = ../../tableau-1.3/build/Release/tabquerycli
TAB_CLI_EXE_LINUX = ../../tableau-1.3/build/Release-x64/tabquerycli
```
The TDVT logs specify that the prams file are not present
```
2022-08-23 13:49:47,223 Thread-[5] CalledProcessError: Return code: 1 ParamFile not specified.

2022-08-23 13:49:47,224 Thread-[6] CalledProcessError: Return code: 1 ParamFile not specified.

2022-08-23 13:49:47,224 Thread-[5] Command line output:ParamFile not specified.

2022-08-23 13:49:47,224 Thread-[6] Command line output:ParamFile not specified.

2022-08-23 13:49:47,225 Thread-[1] CalledProcessError: Return code: 1 ParamFile not specified.
```

Author: Looks like we might need to do some work to add the param files in tdvt_override.ini file and logic to add the command for if tabquerytoolsrv is being used.

Author: Looks like the command that TDVT calls is the same but without the params file 
```
2022-08-23 13:49:45,887 Thread-[6]  calling C:\Program Files\Tableau\Tableau Server\packages\bin.near.22.0802.1408\tabquerytoolsrv.exe --expression-file-list C:\Users\STRIPA~1\AppData\Local\Temp\66c8231c15251d74fd0c4686cda2cd7bf2df7a16c4e0037b69a2a94e3elozhlk\expression_lod_postgres\tests.txt -d D:\dev\near\tableau-tests\tdvt\tds\cast_calcs.postgres.tds --combined --password-file D:\dev\near\tableau-tests\tdvt\tds\postgres.password --output-dir C:\Users\STRIPA~1\AppData\Local\Temp\66c8231c15251d74fd0c4686cda2cd7bf2df7a16c4e0037b69a2a94e3elozhlk -DLogDir=C:\Users\STRIPA~1\AppData\Local\Temp\66c8231c15251d74fd0c4686cda2cd7bf2df7a16c4e0037b69a2a94e3elozhlk\expression_lod_postgres -DOverride=ProtocolServerNewLog -DLogicalQueryRewriteDisable=Funcall:RewriteConstantFuncall -DInMemoryLogicalCacheDisable
```

Reviewer 2: I think that getting this flow working is necessary for TDVT on Linux to be considered done, so the documentation may be premature. May need to schedule more work to update TDVT to do this work for tabquerysrv (while also not breaking our existing Linux testing internally, which uses tabquerytool.) Maybe adding TAB_CLI_LINUX_SRV that will add the params file?

Author: Looks like using CommandLineOverride to add the -DParaFile works. Updating the doc to reflect that.

```
./tabquerytoolsrv  --query-file-list ~/test/tests.txt -d  ~/path_to_tds/cast_calcs.db.tds --combined --password-file ~/path_to_tds/db.password --output-dir ~/test -DLogDir=~/test -DOverride=ProtocolServerNewLog -DLogicalQueryRewriteDisable=Funcall:RewriteConstantFuncall -DInMemoryLogicalCacheDisable -DParamFile="/var/opt/tableau/tableau_server/data/tabsvc/config/minerva_0.<version_code>/workgroup.yml,/var/opt/tableau/tableau_server/data/tabsvc/config/minerva_0.<version_code>/minerva.yml
Attempting to run query...

> Run query successful! Check output file

Reviewer 2: Does this run all of TDVT or just a single test query? The goal of this work is to run TDVT on Linux, having to run each test query xml one by one is not a sufficient solution since they'd need to check that the test passed themselves, and they need to pick the correct input file which is frustrating. Plus, this only works for logical tests, expression tests do not have the inputs in logical query xml.

I haven't looked to much at tabquerysrv, so I may be wrong, but my understanding was that the goal was to have the tdvt.ini file [here](https://github.com/tableau/connector-plugin-sdk/blob/master/tdvt/tdvt/config/tdvt/tdvt.ini) point to tabquerysrv, which would allow users to run TDVT and TDVT would use tabquerysrv under the hood, like it does with tabquerytool for Desktop installations.

Author: Right now adding `tabquerytoolsrv` to `tdvt_override.ini` is not working. Looking into the logs looks like we might need to do some work to add param files in the ini file. I am unaware if we have already done this.

Author: Adding the `-DParamFile` using `CommandLineOverride` worked.

```

Here, 
```
query-file-list: File containing the list of Logical Query files
DLogDir: The directory where logs are written
password-file: Optional path to a file containing the password for the .tds file
DParamFile: Verifies that a valid license exists for tableau server

> ```

Reviewer 2: Not sure what "here" means in this context, thin there's some words missing

Author: Removing `here`

Author: Removed this section since this does not concert the user anymore.

__The agg.countd expression test and the join.null.int logical tests are failing__
Check that your database is correctly returning column nullability information in the metadata. See [Design Considerations]({{ site.baseurl }}/docs/design) for more information.

## Running TDVT test with tabquerytoolsrv in Linux Server

> You can run TDVT tests by configuring the setup file below to use `tabquerytoolsrv`. `tabquerytoolsrv` is located in `/opt/tableau/tableau_server/packages/bin.<version_code>/tabquerytoolsrv`.

Reviewer 2: ```suggestion
You can run TDVT tests by configuring the setup file below to use `tabquerytoolsrv`. `tabquerytoolsrv` is located in `/opt/tableau/tableau_server/packages/bin.<version_code>/tabquerytoolsrv`, where <version_code> corresponds to the version of Tableau Server (ie 2022.3).
```

Took me a moment to figure out what <version_code> meant in this case.

> You can run a particular TDVT tests with the help of a test setup file using `tabquerytoolsrv`. `tabquerytoolsrv` is located in `/opt/tableau/tableau_server/packages/bin.<version_code>/tabquerytoolsrv`.

Reviewer: ```suggestion
You can run TDVT tests by configuring the setup file below to use `tabquerytoolsrv`. `tabquerytoolsrv` is located in `/opt/tableau/tableau_server/packages/bin.<version_code>/tabquerytoolsrv`.
```
